### ISSUE TITLE
Profile events ought to be available as they occur

### ISSUE THREAD

**Issue by lyndonhenry** [souffle-lang/souffle#628]
As of PR #617, if one uses `--profile my-file.log`, profile events are not written to `my-file.log` until the end of computation. Before this, profile events were written to `my-file.log` (approximately) as they occur. 

While dumping only at the end of computation is wonderful in terms of efficiency, many use cases -- for example, when working within runtime or memory constraints -- require that information on events become available as they occur. `souffle-profile` itself even has a "live" mode with the `-l` option to take advantage of this. 

Ideally, `souffle-profile` should be well enough integrated with `souffle` as to not require disk IO as an intermediary, and provide its live information via an interface so that `souffle-profile` can yank it straight from memory. This is no trivial task however, so in the meantime I propose we **give the user the option to dump events as they happen**. 

That's not saying we revert having the ability to dump events only at the end also, that behavior is clearly extremely valuable, but I think we should let users do one or the other by their own choice. And if this switch is not possible to implement as some command line flag, there is always the option of having it as an `--enable-partial-profiling` flag (or whatever) to `./configure`.  

Here is an example, to better explain what I mean. It seems silly at first, but makes more sense as things go on.

~~~
.decl init(x:number) // init = ∅
init(0). // init = { (0) }
init(i + 1) :- init(i), i < 10. // init = { (0) } ∪ { (1), (2), ..., (9) }

.decl loop(x:number) // loop = ∅
.output loop
loop(i) :- init(i). // loop = init
loop(i + 1) :- loop(i). // loop = init ∪ { (10), (11), ..., ∞ }
~~~

Before #617, using `--profile` would give logs for `init`, then also start producing logs for `loop` also. After #617, using `--profile` will give no logs whatsoever. 

Ok, but who cares? The program is stupid, the relation `loop` never halts. Well, unfortunately, as we know this [does tend to happen in general sometimes and there is nothing anyone can do about it](http://www.lel.ed.ac.uk/~gpullum/loopsnoop.html). So partial profiling lets us identify potentially non-halting relations. 

So perhaps it's reasonable to claim that Datalog programs in practice tend to avoid this behavior, I've never come across one myself where the bug was an infinite loop. However, machines are bound by time and memory constraints, and in Datalog programs single relations *are often* responsible for inducing a bottleneck breaking these constraints. It is trivial to envisage a scenario where `loop` has a bounded relation, but that it gets really very big, and crashes our program. Partial profiling information would allow us to see this, however by dumping only at the end, we get no such information whatsoever.

**Comment by yanniss** [comment]
I'd like to second that this is a high importance issue for us as well! One typically doesn't profile things that terminate, only things that don't! Most of the value of the souffle profiling for me has been in the live mode.

**Comment by seanmcl** [comment]
I third!  The profiler has lost essentially all of its value for us.

**Comment by b-scholz** [comment]
The current version is work in progress. I apologise if it is not working as expected. We need more online functionality which we don't have at the moment. 

The current change was necessary because we have high-frequency events (memory usage / CPU usage) and hence we needed to switch to dynamic data-structures rather than writing events continuously to a file. 

To have online profiling, I believe we need to integrate the profiler in souffle, i.e., the performance model is built on the fly from the dynamic data-structure rather than from the file. I envision an integration similar we have for provenance at the moment.

The CLI will permit to write the performance model in text and html format.

**Comment by mmcgr** [comment]
Fixed by #632. Since these changes were noticed so fast, #634, about what profiler changes will be useful, will be of interest.
